In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2
3
5

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

3 6
9


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

ALGORITMO 1

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

4
3 4 5 6
30


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
4
7 12 23 34

Writing input4.txt


In [ ]:
%%writefile output4.txt
782

Writing output4.txt


In [ ]:
%%writefile input5.txt
5
2 4 6 8 10

Writing input5.txt


In [ ]:
%%writefile output5.txt
80

Writing output5.txt


In [ ]:
%%writefile input6.txt
6
15 30 45 60 75 90

Writing input6.txt


In [ ]:
%%writefile output6.txt
6750

Writing output6.txt


In [ ]:
%%writefile input7.txt
7
3 6 9 12 15 18 21

Writing input7.txt


In [ ]:
%%writefile output7.txt
378

Writing output7.txt


In [ ]:
%%writefile input8.txt
8
11 22 33 44 55 66 77 88

Writing input8.txt


In [ ]:
%%writefile output8.txt
6776

Writing output8.txt


In [ ]:
%%writefile input9.txt
9
5 10 15 20 25 30 35 40 45

Writing input9.txt


In [ ]:
%%writefile output9.txt
1800

Writing output9.txt


In [ ]:
%%writefile input10.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input10.txt


In [ ]:
%%writefile output10.txt
90

Writing output10.txt


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > output$i.txt
done

Writing stress_test.sh


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

2147483646

real	5m4.933s
user	5m2.523s
sys	0m0.116s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

2147483646

real	5m5.841s
user	5m3.407s
sys	0m0.131s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

IndentationError: ignored

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10

115



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


115
200000 200000  199999 ... 2 1
^C


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

1/(306/200001000010)

653598039.248366

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.869s
user	0m0.857s
sys	0m0.001s


## Python

In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Writing WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
# Verify
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input14.txt

120989000

real	0m1.531s
user	0m1.460s
sys	0m0.002s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

ALGORITMO 2

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m20.733s
user	2m19.427s
sys	0m0.084s


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {1..10}
do
    ./max_pairwise_product_long_full < input$i.txt > output$i.txt
done

Overwriting stress_test.sh


In [ ]:
print(11 + 7*(200000) + 2*(200000)*(199999) + 1, 11 + 7*(200000) + 2.5*(200000)*(199999))

80001000012 100000900011.0


In [ ]:
# Estimated time by instruction

141/200001000010

7.049964749823752e-10

In [ ]:
# Estimated number of instructions by second

1/(141/200001000010)

1418446808.5815604

In [ ]:
%%writefile max_pairwise_product_n2_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            max_product = max(max_product, l[i] * l[j])

    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_n2_py.py


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

ALGORITMO 3

In [ ]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [ ]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                        #1
    l = list(map(int, input().split()))                     #5n
    l.sort()                                                #nlogn at least
    print(l[n - 1] * l[n - 2])                              #2

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!./max_pairwise_product_sort < input1.txt   #Original array in input1.txt increasing order [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
!./max_pairwise_product_sort < input2.txt   #Original array in input2.txt decreasing order [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

90
90


In [ ]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.004s
user	0m0.001s
sys	0m0.002s


In [ ]:
%%time
 #Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 4.49 ms, sys: 1.24 ms, total: 5.73 ms
Wall time: 104 ms


20 stress tests

In [ ]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [ ]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 17 Product: 850976
n: 25 Product: 857451
n: 43 Product: 980036
n: 10 Product: 692094
n: 30 Product: 839816
n: 48 Product: 956363
n: 9 Product: 475580
n: 46 Product: 954088
n: 44 Product: 902850
n: 17 Product: 981034
n: 40 Product: 928812
n: 7 Product: 267968
n: 42 Product: 991020
n: 27 Product: 908160
n: 23 Product: 990016
n: 2 Product: 141372
n: 4 Product: 418500
n: 11 Product: 557996
n: 42 Product: 963270
n: 17 Product: 690616


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<int> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Writing max_pairwise_product_sort_ins.cpp


In [ ]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [ ]:
!time ./max_pairwise_product_sort_ins < input11.txt

39999800000
Executed instructions: 4321934

real	0m0.133s
user	0m0.126s
sys	0m0.005s


In [ ]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[n - 1] * l[n - 2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")                             #3 + 5n + ins(nlogn timsort) where ins(nlogn sort) are, at least, nlogn.

Writing max_pairwise_product_sort_ins.py


In [ ]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [ ]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 7.79 ms, sys: 1.73 ms, total: 9.52 ms
Wall time: 203 ms


In [ ]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 7.81 ms, sys: 0 ns, total: 7.81 ms
Wall time: 207 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.00683 / 4521931
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.5104166781846075e-09
662068960.4685212


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

ALGORITMO 4

In [ ]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Writing max_pairwise_product_ef.cpp


In [ ]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [ ]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input11.txt

39999800000

real	0m0.070s
user	0m0.069s
sys	0m0.000s


In [ ]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 5.55 ms, sys: 1.09 ms, total: 6.64 ms
Wall time: 105 ms


In [ ]:
%%time
!python max_pairwise_product_ef.py < input11.txt

39999800000
CPU times: user 9.53 ms, sys: 459 µs, total: 9.99 ms
Wall time: 211 ms


20 stress tests

In [ ]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 69 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Writing stress_tests_ef_cpp.cpp


In [ ]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Writing stress_tests_ef_py.py


In [ ]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 36 Product: 863968
n: 57 Product: 980036
n: 60 Product: 898343
n: 31 Product: 927682
n: 13 Product: 762600
n: 51 Product: 954088
n: 12 Product: 683388
n: 26 Product: 981034
n: 70 Product: 960364
n: 14 Product: 858552
n: 61 Product: 993010
n: 52 Product: 963270
n: 10 Product: 631764
n: 32 Product: 994008
n: 21 Product: 960319
n: 24 Product: 828625
n: 2 Product: 74562
n: 54 Product: 904397
n: 8 Product: 886974
n: 23 Product: 911064


In [ ]:
!python stress_tests_ef_py.py

n: 16 Product: 365308852
n: 8 Product: 491973152
n: 75 Product: 582468033
n: 59 Product: 572477246
n: 21 Product: 491749830
n: 42 Product: 576935120
n: 24 Product: 430366608
n: 65 Product: 587932136
n: 88 Product: 588691800
n: 73 Product: 594578880
n: 38 Product: 599441700
n: 40 Product: 515825940
n: 35 Product: 571676974
n: 47 Product: 579124216
n: 62 Product: 577536183
n: 86 Product: 578715162
n: 45 Product: 560760676
n: 31 Product: 550244214
n: 57 Product: 584985970
n: 85 Product: 585610416


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_ef_ins.cpp


In [ ]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Writing max_pairwise_product_ef_ins_py.py


In [ ]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [ ]:
!time ./max_pairwise_product_ef_ins < input11.txt

39999800000
Executed instructions: 2200009

real	0m0.075s
user	0m0.067s
sys	0m0.002s


In [ ]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 8.9 ms, sys: 0 ns, total: 8.9 ms
Wall time: 109 ms


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input11.txt

39999800000
Executed instructions: 2000007
CPU times: user 4.97 ms, sys: 1.95 ms, total: 6.92 ms
Wall time: 206 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.00509/ 2200009
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

2.3136268987990502e-09
432221807.4656189


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

ALGORITMO 5

In [ ]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        if(x > p_max){
            s_max = p_max;
            p_max = x;
        }
        else if(x > s_max){
            s_max = x;
        }
    }

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.003s
user	0m0.000s
sys	0m0.003s


20 stress tests

In [ ]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int x = rand() % 1000 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [ ]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [ ]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [ ]:
!time ./max_pairwise_product_gef_ins_cpp < input11.txt

39999800000
Executed instructions: 1600008

real	0m0.079s
user	0m0.074s
sys	0m0.000s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.073 / 1600008
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.5624771876140615e-08
21917917.80821918


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        i = int(i)
        if(i > p_max):
            s_max = p_max
            p_max = i
        elif i > s_max:
            s_max = i
    print(p_max * s_max)

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 6.77 ms, sys: 0 ns, total: 6.77 ms
Wall time: 105 ms


20 stress tests

In [ ]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(19, 133700)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif i > s_max:
                s_max = i
        print(f"n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_py.py


In [ ]:
!python stress_tests_gef_py.py

n: 93 Product: 17435678260
n: 75 Product: 17729028801
n: 61 Product: 16762952646
n: 99 Product: 17487704500
n: 76 Product: 17019787523
n: 6 Product: 6454362695
n: 56 Product: 17503503186
n: 19 Product: 13787546006
n: 70 Product: 16370434745
n: 28 Product: 17288105470
n: 81 Product: 17602861440
n: 9 Product: 14125847988
n: 75 Product: 16993152000
n: 38 Product: 17283964554
n: 89 Product: 17392834200
n: 61 Product: 17410842338
n: 72 Product: 17206721031
n: 22 Product: 17547691504
n: 67 Product: 17451213821
n: 95 Product: 17175539254


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [ ]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [ ]:
%%time
!python max_pairwise_product_gef_ins_py.py < input11.txt

39999800000
Executed instructions: 1400004
CPU times: user 7.21 ms, sys: 1.03 ms, total: 8.24 ms
Wall time: 306 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.00542 / 1400004
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

3.871417510235685e-09
258303321.0332103


Super eficent Solution Py (T(n)<10n=O(n)))

In [ ]:
%%writefile max_pairwise_product_gef_py.py
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        i = int(i)
        if(i > p_max):
            s_max = p_max
            p_max = i
        elif i > s_max:
            s_max = i
    print(p_max * s_max)

Overwriting max_pairwise_product_gef_py.py


In [ ]:

!python max_pairwise_product_gef_py.py
!python max_pairwise_product_gef_py.py < input_2.txt

2
8 5
40
9300


In [ ]:
%%writefile gen_input.sh
#!/bin/bash

# Número de conjuntos de números aleatorios que deseas generar
numero_conjuntos=100

for ((j=1; j<=$numero_conjuntos; j++)); do
  # Generar un número aleatorio entre 1 y 100
  numero_aleatorio=$((RANDOM % 100 + 1))

  # Crear una cadena con la secuencia de números aleatorios
  numeros_aleatorios=""
  for ((i=1; i<=$numero_aleatorio; i++)); do
    numeros_aleatorios="$numeros_aleatorios $((RANDOM % 100 + 1))"
  done

  # Escribir los datos en un archivo de texto con un nombre único
  archivo_salida="input_$j.txt"
  echo "$numero_aleatorio" > "$archivo_salida"
  echo "$numeros_aleatorios" >> "$archivo_salida"

  echo "Archivo '$archivo_salida' generado con éxito."
done


Overwriting gen_input.sh


In [ ]:
!bash gen_input.sh

Archivo 'input_1.txt' generado con éxito.
Archivo 'input_2.txt' generado con éxito.
Archivo 'input_3.txt' generado con éxito.
Archivo 'input_4.txt' generado con éxito.
Archivo 'input_5.txt' generado con éxito.
Archivo 'input_6.txt' generado con éxito.
Archivo 'input_7.txt' generado con éxito.
Archivo 'input_8.txt' generado con éxito.
Archivo 'input_9.txt' generado con éxito.
Archivo 'input_10.txt' generado con éxito.
Archivo 'input_11.txt' generado con éxito.
Archivo 'input_12.txt' generado con éxito.
Archivo 'input_13.txt' generado con éxito.
Archivo 'input_14.txt' generado con éxito.
Archivo 'input_15.txt' generado con éxito.
Archivo 'input_16.txt' generado con éxito.
Archivo 'input_17.txt' generado con éxito.
Archivo 'input_18.txt' generado con éxito.
Archivo 'input_19.txt' generado con éxito.
Archivo 'input_20.txt' generado con éxito.
Archivo 'input_21.txt' generado con éxito.
Archivo 'input_22.txt' generado con éxito.
Archivo 'input_23.txt' generado con éxito.
Archivo 'input_24.tx

In [ ]:
 %%writefile torture_max_pairwise_sort_super_efficient_python.sh
#!/bin/bash

for i in {0..100}
do
    echo "-----------------------------------"
    echo "Running with input$i.txt"
    python max_pairwise_product_gef_py.py < input_$i.txt

done

Overwriting torture_max_pairwise_sort_super_efficient_python.sh


In [ ]:
%%time
!chmod u+x torture_max_pairwise_sort_super_efficient_python.sh
!./torture_max_pairwise_sort_super_efficient_python.sh

-----------------------------------
Running with input0.txt
./torture_max_pairwise_sort_super_efficient_python.sh: line 7: input_0.txt: No such file or directory
-----------------------------------
Running with input1.txt
5929
-----------------------------------
Running with input2.txt
8835
-----------------------------------
Running with input3.txt
9702
-----------------------------------
Running with input4.txt
4416
-----------------------------------
Running with input5.txt
10000
-----------------------------------
Running with input6.txt
9504
-----------------------------------
Running with input7.txt
-97
-----------------------------------
Running with input8.txt
8280
-----------------------------------
Running with input9.txt
8835
-----------------------------------
Running with input10.txt
7921
-----------------------------------
Running with input11.txt
9408
-----------------------------------
Running with input12.txt
8736
-----------------------------------
Running with input1

#100 stress tests sorting solution in C++ $$(T(n)=O(n \ln g))$$

In [323]:
%%writefile stress_tests_hundredcases_sort_cpp.cpp

#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>

int main() {

    std::srand(std::time(nullptr));

    for (int i = 0; i < 100; i++) {
        int n = rand() % 100 + 1;
        std::vector<int> v(n);

        for (int j = 0; j < n; j++) {
            v[j] = rand() % 20000 + 1;
        }

        std::sort(v.begin(), v.end());

        long long product = static_cast<long long>(v[n - 1]) * v[n - 2];

        std::cout << "Input: " << i + 1 << " n: " << n << " Resultado: " << product << '\n';
    }

    return 0;

}



Overwriting stress_tests_hundredcases_sort_cpp.cpp


In [325]:
%%time
!g++ stress_tests_hundredcases_sort_cpp.cpp -o hundred_sort_cpp
!./hundred_sort_cpp

Input: 1 n: 34 Resultado: 374382070
Input: 2 n: 62 Resultado: 390614868
Input: 3 n: 52 Resultado: 394631284
Input: 4 n: 96 Resultado: 396208856
Input: 5 n: 46 Resultado: 389862816
Input: 6 n: 3 Resultado: 250264984
Input: 7 n: 50 Resultado: 377893950
Input: 8 n: 40 Resultado: 359552115
Input: 9 n: 98 Resultado: 391089729
Input: 10 n: 85 Resultado: 395134595
Input: 11 n: 37 Resultado: 378470880
Input: 12 n: 72 Resultado: 393211739
Input: 13 n: 62 Resultado: 388740330
Input: 14 n: 97 Resultado: 382456368
Input: 15 n: 9 Resultado: 369273729
Input: 16 n: 94 Resultado: 399240312
Input: 17 n: 7 Resultado: 365133650
Input: 18 n: 56 Resultado: 395491473
Input: 19 n: 82 Resultado: 396106644
Input: 20 n: 69 Resultado: 381242798
Input: 21 n: 73 Resultado: 373987020
Input: 22 n: 55 Resultado: 387460104
Input: 23 n: 76 Resultado: 392890932
Input: 24 n: 8 Resultado: 145589715
Input: 25 n: 31 Resultado: 391739864
Input: 26 n: 71 Resultado: 362319360
Input: 27 n: 80 Resultado: 386077425
Input: 28 n: 3

In [326]:
%%writefile gen_input_sort_cpp.sh
#!/bin/bash

numero_conjuntos=100

for ((j=1; j<=$numero_conjuntos; j++)); do
  numero_aleatorio=$((RANDOM % 100 + 1))
  numeros_aleatorios=""
  for ((i=1; i<=$numero_aleatorio; i++)); do
    numeros_aleatorios="$numeros_aleatorios $((RANDOM % 100 + 1))"
  done

  archivo_salida="input_$j.txt"
  echo "$numero_aleatorio" > "$archivo_salida"
  echo "$numeros_aleatorios" >> "$archivo_salida"

  echo "Archivo '$archivo_salida' generado con éxito."
done

Overwriting gen_input_sort_cpp.sh


In [327]:
!bash gen_input_sort_cpp.sh

Archivo 'input_1.txt' generado con éxito.
Archivo 'input_2.txt' generado con éxito.
Archivo 'input_3.txt' generado con éxito.
Archivo 'input_4.txt' generado con éxito.
Archivo 'input_5.txt' generado con éxito.
Archivo 'input_6.txt' generado con éxito.
Archivo 'input_7.txt' generado con éxito.
Archivo 'input_8.txt' generado con éxito.
Archivo 'input_9.txt' generado con éxito.
Archivo 'input_10.txt' generado con éxito.
Archivo 'input_11.txt' generado con éxito.
Archivo 'input_12.txt' generado con éxito.
Archivo 'input_13.txt' generado con éxito.
Archivo 'input_14.txt' generado con éxito.
Archivo 'input_15.txt' generado con éxito.
Archivo 'input_16.txt' generado con éxito.
Archivo 'input_17.txt' generado con éxito.
Archivo 'input_18.txt' generado con éxito.
Archivo 'input_19.txt' generado con éxito.
Archivo 'input_20.txt' generado con éxito.
Archivo 'input_21.txt' generado con éxito.
Archivo 'input_22.txt' generado con éxito.
Archivo 'input_23.txt' generado con éxito.
Archivo 'input_24.tx

In [329]:
%%writefile torture_max_pairwise_sort_solution_cpp.sh
#!/bin/bash

for i in {1..100}
do
    echo "-----------------------------------"
    echo "Running with input$i.txt"
    ./hundred_sort_cpp < input_$i.txt

done

Overwriting torture_max_pairwise_sort_solution_cpp.sh


In [330]:
%%time
!chmod u+x torture_max_pairwise_sort_solution_cpp.sh
!./torture_max_pairwise_sort_solution_cpp.sh

Streaming output truncated to the last 5000 lines.
-----------------------------------
Running with input52.txt
Input: 1 n: 97 Resultado: 387771143
Input: 2 n: 15 Resultado: 321915825
Input: 3 n: 17 Resultado: 350346304
Input: 4 n: 49 Resultado: 368206830
Input: 5 n: 87 Resultado: 393823925
Input: 6 n: 31 Resultado: 389161374
Input: 7 n: 31 Resultado: 368509433
Input: 8 n: 91 Resultado: 391822218
Input: 9 n: 99 Resultado: 395505000
Input: 10 n: 74 Resultado: 392848188
Input: 11 n: 89 Resultado: 391296899
Input: 12 n: 84 Resultado: 396945546
Input: 13 n: 73 Resultado: 386241972
Input: 14 n: 64 Resultado: 399640077
Input: 15 n: 10 Resultado: 299585152
Input: 16 n: 85 Resultado: 373100287
Input: 17 n: 8 Resultado: 355714902
Input: 18 n: 60 Resultado: 381941586
Input: 19 n: 87 Resultado: 385183408
Input: 20 n: 99 Resultado: 390397632
Input: 21 n: 76 Resultado: 354132862
Input: 22 n: 37 Resultado: 339452178
Input: 23 n: 18 Resultado: 319868703
Input: 24 n: 97 Resultado: 396188700
Input: 25 

#100 stress tests super efficient solution in C++ $$ (T(n)<10 n=O(n)) $$

In [317]:
%%writefile stress_tests_hundredcases_cpp.cpp

#include <iostream>
#include <ctime>
#include <cstdlib>
#include <vector>
#include <climits>

int main() {

    std::srand(static_cast<unsigned int>(std::time(nullptr)));

    for (int i = 0; i < 100; i++) {
        int n = std::rand() % 100 + 1;
        int64_t a_max = INT_MIN;
        int64_t b_max = INT_MIN;

        for (int j = 0; j < n; j++) {
            int64_t x = std::rand() % 20000 + 1;
            if (x > a_max) {
                b_max = a_max;
                a_max = x;
            } else if (x > b_max) {
                b_max = x;
            }
        }

        std::cout << "Input: " << i + 1 << " n: " << n << " Resultado: " << a_max * b_max << "\n";
    }

    return 0;
}



Overwriting stress_tests_hundredcases_cpp.cpp


In [318]:
%%time
!g++ stress_tests_hundredcases_cpp.cpp -o hundred_cpp
!./hundred_cpp

Input: 1 n: 77 Resultado: 385371672
Input: 2 n: 29 Resultado: 357199438
Input: 3 n: 41 Resultado: 380192494
Input: 4 n: 23 Resultado: 379558300
Input: 5 n: 93 Resultado: 391584660
Input: 6 n: 88 Resultado: 396325328
Input: 7 n: 12 Resultado: 330532160
Input: 8 n: 30 Resultado: 383937788
Input: 9 n: 71 Resultado: 395030098
Input: 10 n: 79 Resultado: 392930174
Input: 11 n: 52 Resultado: 390636181
Input: 12 n: 42 Resultado: 393861012
Input: 13 n: 12 Resultado: 344002274
Input: 14 n: 78 Resultado: 380904264
Input: 15 n: 63 Resultado: 395680000
Input: 16 n: 93 Resultado: 375628435
Input: 17 n: 22 Resultado: 385824450
Input: 18 n: 97 Resultado: 387157760
Input: 19 n: 66 Resultado: 398241360
Input: 20 n: 4 Resultado: 248041630
Input: 21 n: 30 Resultado: 373933336
Input: 22 n: 95 Resultado: 370341078
Input: 23 n: 1 Resultado: -36077725286400
Input: 24 n: 24 Resultado: 385726236
Input: 25 n: 28 Resultado: 373378304
Input: 26 n: 30 Resultado: 387903790
Input: 27 n: 41 Resultado: 370160439
Input:

In [319]:
%%writefile gen_input_cpp.sh
#!/bin/bash

numero_conjuntos=100

for ((j=1; j<=$numero_conjuntos; j++)); do
  numero_aleatorio=$((RANDOM % 100 + 1))
  numeros_aleatorios=""
  for ((i=1; i<=$numero_aleatorio; i++)); do
    numeros_aleatorios="$numeros_aleatorios $((RANDOM % 100 + 1))"
  done

  archivo_salida="input_$j.txt"
  echo "$numero_aleatorio" > "$archivo_salida"
  echo "$numeros_aleatorios" >> "$archivo_salida"

  echo "Archivo '$archivo_salida' generado con éxito."
done

Overwriting gen_input_cpp.sh


In [320]:
!bash gen_input_cpp.sh

Archivo 'input_1.txt' generado con éxito.
Archivo 'input_2.txt' generado con éxito.
Archivo 'input_3.txt' generado con éxito.
Archivo 'input_4.txt' generado con éxito.
Archivo 'input_5.txt' generado con éxito.
Archivo 'input_6.txt' generado con éxito.
Archivo 'input_7.txt' generado con éxito.
Archivo 'input_8.txt' generado con éxito.
Archivo 'input_9.txt' generado con éxito.
Archivo 'input_10.txt' generado con éxito.
Archivo 'input_11.txt' generado con éxito.
Archivo 'input_12.txt' generado con éxito.
Archivo 'input_13.txt' generado con éxito.
Archivo 'input_14.txt' generado con éxito.
Archivo 'input_15.txt' generado con éxito.
Archivo 'input_16.txt' generado con éxito.
Archivo 'input_17.txt' generado con éxito.
Archivo 'input_18.txt' generado con éxito.
Archivo 'input_19.txt' generado con éxito.
Archivo 'input_20.txt' generado con éxito.
Archivo 'input_21.txt' generado con éxito.
Archivo 'input_22.txt' generado con éxito.
Archivo 'input_23.txt' generado con éxito.
Archivo 'input_24.tx

In [321]:
 %%writefile torture_max_pairwise_sort_super_efficient_cpp.sh
#!/bin/bash


for i in {1..100}
do
    echo "-----------------------------------"
    echo "Running with input$i.txt"
    ./stress_tests_hundredcases_cpp < input_$i.txt > input_$i.txt

done

Overwriting torture_max_pairwise_sort_super_efficient_cpp.sh


In [322]:
%%time
!chmod u+x torture_max_pairwise_sort_super_efficient_cpp.sh
!./torture_max_pairwise_sort_super_efficient_cpp.sh

-----------------------------------
Running with input1.txt
-----------------------------------
Running with input2.txt
-----------------------------------
Running with input3.txt
-----------------------------------
Running with input4.txt
-----------------------------------
Running with input5.txt
-----------------------------------
Running with input6.txt
-----------------------------------
Running with input7.txt
-----------------------------------
Running with input8.txt
-----------------------------------
Running with input9.txt
-----------------------------------
Running with input10.txt
-----------------------------------
Running with input11.txt
-----------------------------------
Running with input12.txt
-----------------------------------
Running with input13.txt
-----------------------------------
Running with input14.txt
-----------------------------------
Running with input15.txt
-----------------------------------
Running with input16.txt
---------------------------------